Half Donut

curran opened the pull request:
This PR introduces support for a "half donut" chart.

<img width="1174" alt="image" src="https://github.com/user-attachments/assets/e979be46-e783-4be4-b416-6a2c7244d8a1">

<img width="1170" alt="image" src="https://github.com/user-attachments/assets/30c47a78-5605-4616-8151-cd3bdc63927d">

<img width="1191" alt="image" src="https://github.com/user-attachments/assets/ab25f279-d7ec-4c26-be7e-c336f1555ce7">

<img width="1191" alt="image" src="https://github.com/user-attachments/assets/f3479a18-cfb6-4036-935a-cb0cace701ef">

Summary of changes:

 * Export constants that define half donut angle ranges
 * Add a new example that cycles through orientations (top, right, bottom, left) and also lets you resize dynamically  (height > width, height < width)
 * Make the scaling work for all 8 cases (top, right, bottom, left) X (height > width, height < width)

Based on a previous attempt in

 * https://github.com/f5/unovis/pull/444/files

curran commented:
Current status: distracted by TypeScript linting warning:

<img width="824" alt="image" src="https://github.com/user-attachments/assets/695ba13c-a4ff-4f45-93f6-ae30e5cfa7e9">

I thought it would pick up the change... Maybe I need to build or something?

curran commented:
Got it - just needed to run `npm run build:ts`

<img width="806" alt="image" src="https://github.com/user-attachments/assets/401bb3a3-a085-4605-afd2-f1f340d0e539">

curran commented:
Current status:

<img width="1019" alt="image" src="https://github.com/user-attachments/assets/98b624c8-9529-4e1e-9a70-525b94e2cf53">


 * Solved left and right cases
 * First pass of the feature is ready
 * This will require custom margins to be passed in from the outside in the case that there is a text label
 * IMO this is acceptable, since a more automatic solution would probably involve additional complexity around measuring text bounding boxes, which may not be worth it

curran commented on packages/solid/licences.txt:
```diff
@@ -5,7 +5,7 @@ rollup               perpetual       MIT           4.24.2             Rich Harri
 tsx                  perpetual       MIT           4.19.2             Hiroki Osame hiroki.osame@gmail.com
 typescript           perpetual       Apache-2.0    5.6.3              Microsoft Corp.
-vite                 perpetual       MIT           5.4.10             Evan You
+vite                 perpetual       MIT           5.4.11             Evan You
```
```suggestion
vite                 perpetual       MIT           5.4.10             Evan You
```

krisexa commented:
Looks good @curran -- could be the screenshots, but can you confirm that the space between the sections is the same as the other pie charts

<img width="707" alt="Screenshot 2024-12-05 at 1 26 06 PM" src="https://github.com/user-attachments/assets/203f1785-03ef-440d-ae78-0a74a394e77c">

rokotyan commented on packages/ts/src/components/donut/index.ts:
```diff
@@ -68,9 +81,47 @@ export class Donut<Datum> extends ComponentCore<Datum[], DonutConfigInterface<Da
+        : this._height / 2
+
+    const translateX = isHalfDonutLeft
+      ? this._width - Math.abs(this._width - outerRadius) / 2
```
Same here, or am I missing something?

curran replied:
Great catch! Simplified.

curran commented:
The size of the gap in the middle depends on the size of the container.

rokotyan reviewed (commented):
Great job! Left you a few comments.

Also:
- We probably shouldn't merge to `main` if we want to do regular updates from the `upstream`. I'll think about the workflow for this.
- Unovis doesn't squash the commits so we should clean up the commit history to have only two: one for the component changes (`Component` commitlint prefix), and another one for the dev example (`Dev` commitlint prefix).
- We should see how to raise PRs to the main repo

rokotyan commented:
@krisexa Are you talking about this?
<img width="142" alt="image" src="https://github.com/user-attachments/assets/f85cd572-8f4f-4f35-89d1-e4b96953b391">

If so, you can just ignore it.

krisexa commented:
Yeah AOK then

curran commented on packages/ts/src/components/donut/index.ts:
```diff
@@ -25,6 +25,19 @@ import { createArc, updateArc, removeArc } from './modules/arc'
 import * as s from './style'
 
+// Constants that support half donuts
+export const halfDonutAngleRanges = Array.from({ length: 4 }, (_, i): [number, number] => {
```
Done.

curran commented on packages/ts/src/components/donut/index.ts:
```diff
@@ -68,9 +81,47 @@ export class Donut<Datum> extends ComponentCore<Datum[], DonutConfigInterface<Da
     const innerRadius = config.arcWidth === 0 ? 0 : clamp(outerRadius - config.arcWidth, 0, outerRadius - 1)
 
+    const translateY = isHalfDonutTop
+      ? this._height - Math.abs(this._height - outerRadius) / 2
```
Great catch! Simplified.

curran commented:
@rokotyan This is ready for re-review.

Summary of changes since last review:

 * Addressed code feedback
 * Changed the base of the PR to `[1.5.0-exaforce](https://github.com/ExaForce/unovis/tree/1.5.0-exaforce)`
 * Also opened a PR with the same changes in https://github.com/f5/unovis/pull/502

I suppose we can merge this one here and I'll let you take it from there re: package publishing. Cheers!

curran commented:
Investigating how to do this part:

>Unovis doesn't squash the commits so we should clean up the commit history to have only two: one for the component changes (Component commitlint prefix), and another one for the dev example (Dev commitlint prefix).

curran commented:
The additional work around rearranging the changes into two commits, one for the component changes and one for the dev example, seems to be fairly involved, and has some downsides to it as well. For example, we'll lose the thread here of the feedback comments interleaved with the commits, as discussed in [StackOverflow: [How to squash commits in git after they have been pushed?](https://stackoverflow.com/questions/5667884/how-to-squash-commits-in-git-after-they-have-been-pushed)](https://stackoverflow.com/questions/5667884/how-to-squash-commits-in-git-after-they-have-been-pushed).

Would it be acceptable to squash the commits into just one instead of two? That seems more readily doable with off-the-shelf git commands. For making two separate commits, I'm not sure how to go about that. I would need to somehow "un-stage" all the changes that have already been committed and pushed, stage one file, commit that, stage the other file, then commit that. I'll see what I can do but it doesn't seem so straightforward. If you have any suggestions for the specific commands to use to make this happen, any guidance would be appreciated.

I'm also not sure who this is for... Is there someone who cares deeply about the cleanliness of the Git commit history? Is that a philosophy of the project or something? I don't understand why that matters.

For the long term, it would be great if we could enable commit squashing at the GitHub level, so we don't need to spend additional time thinking about how to clean up the git history. Since we're on a fork, maybe that's possible?

Will continue trying and see what I can do... The instructions here look promising: [Reddit: Is there a way to remove all local commits while keeping the changes?](https://www.reddit.com/r/git/comments/sum03x/is_there_a_way_to_remove_all_local_commits_while/). In particular:

```
git reset origin/main
git clean -fd
git restore .
git add packages/ts/src/components/donut
git commit -m "Component | Donut: Half Donut"
git add packages/dev/src/examples/misc/donut/half-donut-full-height/index.tsx
git commit -m "Dev | Example: Half Donut"
git push -f
```

or something like that... This is uncharted territory for me and I fear I might mess up the branch. Will give it a try!

rokotyan commented:
@curran I do `git reset --soft` on a regular basis to clean up my commit history when needed. You might need to plan certain things in advance if you're adding multiple changes to the same files and want to have a separate commit history. In this specific case, it should be rather simple as the change is relatively small.

rokotyan commented:
> For the long term, it would be great if we could enable commit squashing at the GitHub level, so we don't need to spend additional time thinking about how to clean up the git history. Since we're on a fork, maybe that's possible?

Since we're planning to raise PRs to the main repo from the same branches, I don't think it'll solve the problem. And I always appreciate having a clean and structured git history — it's helpful when you need to find certain things in it.

devgru commented:
@curran From my experience, the easiest way to comb the history is git's `interactive rebase`, especially `f` and `e` commands to squash comments and update their messages.

curran commented:
Thanks for the tip! I started looking into that, but I don't think that would work for this case since we don't want to just squash groups of commits, but rather group files together into commits, aggregating all the changes from the whole thread of commits.

This is what finally worked:

```
git reset --soft HEAD~9
git reset # Required to un-stage the changes
git add packages/ts/src/components.ts
git add packages/ts/src/components/donut/index.ts
git commit -m "Component | Donut: Half donut support"
git status -s
git add . # stage the new example
git commit -m "Dev | Donut: Half donut example"
git push --force
```

rokotyan reviewed (approved):
Appvd
